numpy/__init__.pyi

```diff
@@ -1502,8 +1502,8 @@ _FlexDType = TypeVar("_FlexDType", bound=dtype[flexible])
 
 # TODO: Set the `bound` to something more suitable once we
 # have proper shape support
-_ShapeType = TypeVar("_ShapeType", bound=Any)
-_ShapeType2 = TypeVar("_ShapeType2", bound=Any)
+_ShapeType = TypeVar("_ShapeType", covariant=True, bound=tuple[int, ...])
+_ShapeType2 = TypeVar("_ShapeType2", covariant=True, bound=tuple[int, ...])
```

Comment:
I'm confused by this one - I'm happy to make the change (`__array_wrap__` is sort of deprecated), but if it's used as a shape type, why wouldn't it be covariant?  And if, as you say, it's not bound to any type, why keep the bound?

Reply:
When a type parameter is co/contra-variant, it is always with respect to a generic *type*.
For generic functions, there is no such thing as a covariant type parameter; there's nothing to vary.

Taking function parameters as analogue, it'd be a bit like taking the keyword-only `out` parameter of a specific `ufunc`, and re-using it as a method parameter of some unrelated type. 

Both function parameters and type parameters can be seen as (contextual) placeholders.
But where function parameters accept *values* or *instances*, type parameters accept only, and suprisingly, *types* (plus they live in the typing-dimension, as opposed to the runtime one, and they're interconnected with wormholes, but I digress).

BTW, that's one of the reasons why I often use Python 3.12 syntax when giving typing examples, as it doesn't have these problems, and the variance of type params is automagically inferred (even though it can't always read your mind it turns out, but let's not go there now).

Reply:
I don't understand all of that, but it definitely helps me get partway through the wormhole.  I'm happy to make the change.  And yeah, python 3.12 syntax does make variable scope much clearer :)